libraries/chain/controller.cpp

```diff
@@ -4058,26 +4043,59 @@ struct controller_impl {
       }
    }
 
-   // thread safe, expected to be called from thread other than the main thread
-   template<typename ForkDB, typename BS>
-   block_handle create_block_state_i( ForkDB& forkdb, const block_id_type& id, const signed_block_ptr& b, const BS& prev ) {
+   // verify basic_block invariants
+   template<typename BS>
+   std::optional<qc_t> verify_basic_block_invariants(const signed_block_ptr& b, const BS& prev) {
       constexpr bool is_proper_savanna_block = std::is_same_v<typename std::decay_t<BS>, block_state>;
       assert(is_proper_savanna_block == b->is_proper_svnn_block());
 
       if constexpr (is_proper_savanna_block) {
          EOS_ASSERT( b->is_proper_svnn_block(), block_validate_exception,
                      "create_block_state_i cannot be called on block #${b} which is not a Proper Savanna block unless the prev block state provided is of type block_state",
                      ("b", b->block_num()) );
-         verify_proper_block_exts(id, b, prev);
+         return verify_basic_proper_block_invariants(b, prev);
       } else {
          EOS_ASSERT( !b->is_proper_svnn_block(), block_validate_exception,
                      "create_block_state_i cannot be called on block #${b} which is a Proper Savanna block unless the prev block state provided is of type block_state_legacy",
                      ("b", b->block_num()) );
 
          if (b->is_legacy_block()) {
-            verify_legacy_block_exts(b, prev);
+            verify_legacy_block_invariants(b, prev);
          } else {
-            verify_transition_block_exts(b, prev);
+            verify_transition_block_invariants(b, prev);
+         }
+      }
+
+      return {};
+   }
+
+   // This verifies BLS signatures and is expensive.
+   void verify_qc( const signed_block_ptr& b, const block_header_state& prev, const qc_t& qc ) {
+      // find the claimed block's block state on branch of id
+      auto bsp = fork_db_fetch_bsp_on_branch_by_num( prev.id(), qc.block_num );
+      EOS_ASSERT( bsp, invalid_qc_claim,
+                  "Block state was not found in forkdb for claimed block ${bn}. Current block number: ${b}",
+                  ("bn", qc.block_num)("b", b->block_num()) );
+
+      bsp->verify_qc(qc);
+   }
+
+   // thread safe, expected to be called from thread other than the main thread
+   template<typename ForkDB, typename BS>
+   block_handle create_block_state_i( ForkDB& forkdb, const block_id_type& id, const signed_block_ptr& b, const BS& prev ) {
+      constexpr bool is_proper_savanna_block = std::is_same_v<typename std::decay_t<BS>, block_state>;
+      assert(is_proper_savanna_block == b->is_proper_svnn_block());
+
+      std::optional<qc_t> qc = verify_basic_block_invariants(b, prev);
+
+      if constexpr (is_proper_savanna_block) {
+         if (qc) {
+            verify_qc(b, prev, *qc);
+
+            const auto  qc_claim = qc->to_qc_claim();
```

Comment:
In the followup PR, will remove this statement and move `qc->to_qc_claim()` into the `dlog` below so it is not executed if `dlog` is not enabled

Reply:
https://github.com/AntelopeIO/spring/pull/735